### PR TITLE
Refactored DateAndTime to remove POCO and Logging Dependencies

### LINKED ISSUE #20523: Remove logging and POCO dependencies from DateAndTime
This issue is related to removing logger and poco dependencies from `Mantid::Kernel::DateAndTime` in preparation for moving the module out of Kernel into a lower level library for use elsewhere.

### PR DESCRIPTION
**Pull request by LamarMoore**
Logging and poco dependencies were removed from `DateAndTime` in preparation for moving this library to the lower level `Mantid::Types` library.

**To test:**

Code review and ensure unit tests pass. Would be good to have input on any edge cases I may have missed regarding file loading and date/time formats.

Fixes #20523. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
@LamarMoore Could you avoid refactoring tons of Doxygen with every PR? It makes the diff hard to read. Maybe this is a setting issue in your IDE?

Could you undo the formatting changes here?

File: Framework/Kernel/src/DateAndTime.cpp
Comment (SimonHeybrock) [inline Framework/Kernel/src/DateAndTime.cpp]: Does is make sense to add a unit test, checking that this throws now?

**Comment by LamarMoore** [comment]
@SimonHeybrock It is probably clang on my system which reformatting the doxygen. I will look into it.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Looks good, but please undo the reformatting of all the doxygen to keep the diff reasonable and avoid merge conflicts.

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Can we capitalize this for consistency?

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Should this in fact be in the helpers namespace?

File: Framework/Kernel/src/DateAndTime.cpp
Comment (martyngigg) [inline Framework/Kernel/src/DateAndTime.cpp]: --> `<cmath>`

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Can we convert these to the C++ versions: `<cmath>` & `<cstdint>`?

File: Framework/Kernel/src/DateAndTime.cpp
Comment (martyngigg) [reply Framework/Kernel/src/DateAndTime.cpp]: Agreed.

File: Framework/Kernel/test/DateAndTimeTest.h
Comment (martyngigg) [inline Framework/Kernel/test/DateAndTimeTest.h]: Do we need `cout` in the test?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Generally looks good. Just a few minor comments

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (LamarMoore) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: It is used internally (within `DateAndTime`) as well.

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Should this be `<ctime>`?

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Is there a reason we have one defined in the header and the others in the `.cpp` file?

File: Framework/Kernel/src/DateAndTime.cpp
Comment (martyngigg) [inline Framework/Kernel/src/DateAndTime.cpp]: We could change the signature of `setFromISO8601` to `setFromISO8601(std::string time)` and remove the manual copy? That way we let the compiler do a move if possible: https://stackoverflow.com/questions/46028709/why-is-passing-by-value-if-a-copy-is-needed-recommended-in-c11-if-a-const-re

File: Framework/Kernel/test/DateAndTimeTest.h
Comment (martyngigg) [inline Framework/Kernel/test/DateAndTimeTest.h]: `<ctime>`

File: Framework/Kernel/test/DateAndTimeTest.h
Comment (martyngigg) [inline Framework/Kernel/test/DateAndTimeTest.h]: Could we add some tests with strings that look a bit like a date time, i.e the old strings that would have passed but will now fail?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (LamarMoore) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Yes, we cannot define non-integral types in the header. Builds fail.

File: Framework/Kernel/src/DateAndTime.cpp
Comment (LamarMoore) [reply Framework/Kernel/src/DateAndTime.cpp]: `str` is used within the exception handling (which is why it is preserved)

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (LamarMoore) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: @martyngigg I do not think we can initialise non-integral types in a header without some compilers complaining.

File: Framework/Kernel/src/DateAndTime.cpp
Comment (martyngigg) [reply Framework/Kernel/src/DateAndTime.cpp]: Oh yes, I missed that. In that case we should stick with the original `const &` version to avoid having 2 copies when an l-value input is used.

Either that or we remove the `str` from the exception and just throw the `std::invalid_argument` without including the string. The caller could then decide whether to include it in their error message if they want. It seems a shame that most of the time when the string is valid we are paying the price of a copy just for the error message.

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h
Comment (martyngigg) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Yes, static non-integral types have to be defined out of line. I was suggesting that we be consistent and put `EPOCH_DIFF` in `.cpp` too. I'm not overly concerned though.

**Review by martyngigg — APPROVED** [review]
I'm happy with this to be merged now.

**Comment by SimonHeybrock** [comment]
@LamarMoore The diff is showing unrelated changes from our `LoadEventNexus` feature branch? If you remove those commits I could merge this.

**Comment by SimonHeybrock** [comment]
@LamarMoore Removed unrelated commits as dicussed on slack, looks ok, and github show previous review comments. 👍

**Review by SimonHeybrock — APPROVED** [review]

File: Framework/Kernel/src/DateAndTimeHelpers.cpp
Comment (OwenArnold) [inline Framework/Kernel/src/DateAndTimeHelpers.cpp]: For future readers, i think it would be useful to have something in the comments here, or for `isArgusDateTime` which gives and example of what that date-time string is like. and what the corresponding sanitized string ends up like. I wouldn't object if this was done as a separate PR, as I know that this PR is holding up others.

File: Framework/Kernel/test/DateAndTimeHelpersTest.h
Comment (OwenArnold) [inline Framework/Kernel/test/DateAndTimeHelpersTest.h]: As a follow on from my comment above. is there any reason that the other methods in this namespace can't be tested? I think that would be useful.

File: Framework/Kernel/src/DateAndTimeHelpers.cpp
Comment (LamarMoore) [reply Framework/Kernel/src/DateAndTimeHelpers.cpp]: This will be added in this PR as changes are made there for DateAndTime anyway https://github.com/mantidproject/mantid/pull/20481

File: Framework/Kernel/test/DateAndTimeHelpersTest.h
Comment (LamarMoore) [reply Framework/Kernel/test/DateAndTimeHelpersTest.h]: Same as above. Will be added in this PR https://github.com/mantidproject/mantid/pull/20481

**Comment by martyngigg** [comment]
@OwenArnold I'll merge this once the builds are happy

**Pull request merged**
